File: library/src/test/java/me/proxer/library/api/BooleanAdapterFactoryTest.java
package me.proxer.library.api;

import com.squareup.moshi.JsonAdapter;
import com.squareup.moshi.Moshi;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

import java.lang.annotation.Annotation;
import java.util.Collections;
import java.util.HashSet;

import static org.assertj.core.api.Assertions.assertThat;

/**
 * @author Ruben Gees
 */
class BooleanAdapterFactoryTest {

    private BooleanAdapterFactory factory;
    private Moshi moshi;

    private Annotation numberBasedBooleanAnnotation;

    @BeforeEach
    void setUp() {
        factory = new BooleanAdapterFactory();
        moshi = new Moshi.Builder().add(factory).build();

        numberBasedBooleanAnnotation = Dummy.class.getAnnotation(NumberBasedBoolean.class);
    }

    @Test
    void testCreatePrimitive() {
        final JsonAdapter<?> defaultBooleanAdapter = new Moshi.Builder().build().adapter(boolean.class);

        assertThat(factory.create(boolean.class, Collections.emptySet(), moshi))
                .isNotNull()
                .isInstanceOf(defaultBooleanAdapter.getClass());
    }

    @Test
    void testCreateBoxed() {
        final JsonAdapter<?> defaultBooleanAdapter = new Moshi.Builder().build().adapter(Boolean.class);

        assertThat(factory.create(Boolean.class, Collections.emptySet(), moshi))
                .isNotNull()
                .isInstanceOf(defaultBooleanAdapter.getClass());
    }

    @Test
    void testCreatePrimitiveNumberBased() {
        final JsonAdapter<?> defaultBooleanAdapter = new Moshi.Builder().build().adapter(boolean.class);

        assertThat(factory.create(boolean.class, Collections.singleton(numberBasedBooleanAnnotation), moshi))
                .isNotNull()
                .isNotInstanceOf(defaultBooleanAdapter.getClass());
    }

    @Test
    void testCreateBoxedNumberBased() {
        final JsonAdapter<?> defaultBooleanAdapter = new Moshi.Builder().build().adapter(Boolean.class);

        assertThat(factory.create(Boolean.class, Collections.singleton(numberBasedBooleanAnnotation), moshi))
                .isNotNull()
                .isNotInstanceOf(defaultBooleanAdapter.getClass());
    }

    @Test
    void testCreateNonBoolean() {
        assertThat(factory.create(String.class, new HashSet<>(), moshi)).isNull();
    }

    @NumberBasedBoolean
    private static class Dummy {
    }
}

File: library/src/test/kotlin/me/proxer/library/api/BooleanAdapterFactoryTest.kt
package me.proxer.library.api

import com.squareup.moshi.Moshi
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Test
import java.util.HashSet

/**
 * @author Ruben Gees
 */
class BooleanAdapterFactoryTest {

    private val factory = BooleanAdapterFactory()
    private val moshi = Moshi.Builder().add(factory).build()

    private val numberBasedBooleanAnnotation = Dummy::class.java.getAnnotation(NumberBasedBoolean::class.java)

    private val primitiveBoolean = Boolean::class.javaPrimitiveType as Class<Boolean>
    private val boxedBoolean = Boolean::class.java

    @Test
    fun testCreatePrimitive() {
        val defaultBooleanAdapter = Moshi.Builder().build().adapter(primitiveBoolean)

        assertThat(factory.create(primitiveBoolean, emptySet(), moshi))
            .isNotNull
            .isInstanceOf(defaultBooleanAdapter.javaClass)
    }

    @Test
    fun testCreateBoxed() {
        val defaultBooleanAdapter = Moshi.Builder().build().adapter(boxedBoolean)

        assertThat(factory.create(Boolean::class.java, emptySet(), moshi))
            .isNotNull
            .isInstanceOf(defaultBooleanAdapter.javaClass)
    }

    @Test
    fun testCreatePrimitiveNumberBased() {
        val defaultBooleanAdapter = Moshi.Builder().build().adapter(primitiveBoolean)

        assertThat(factory.create(primitiveBoolean, setOf<Annotation>(numberBasedBooleanAnnotation), moshi))
            .isNotNull
            .isNotInstanceOf(defaultBooleanAdapter.javaClass)
    }

    @Test
    fun testCreateBoxedNumberBased() {
        val defaultBooleanAdapter = Moshi.Builder().build().adapter(boxedBoolean)

        assertThat(factory.create(boxedBoolean, setOf<Annotation>(numberBasedBooleanAnnotation), moshi))
            .isNotNull
            .isNotInstanceOf(defaultBooleanAdapter.javaClass)
    }

    @Test
    fun testCreateNonBoolean() {
        assertThat(factory.create(String::class.java, HashSet(), moshi)).isNull()
    }

    @NumberBasedBoolean
    private class Dummy
}
